### DISCUSSION TITLE
How do you document use cases?

### DISCUSSION THREAD

**Discussion by pawelpodsiadly** [kgrzybek/modular-monolith-with-ddd#180] [Q&A] [answered]
Hi,

I have been wondering what would be the best approach to document use cases? Do you document all use cases using event storming notation?

I find it hard to track what a given use case consist of when it looks for example like: 
CommandHandler -> Aggregate method -> Domain Event -> DomainEvent Handler -> Domain Event -> Domain Event Handler -> Integration Event etc

Thanks

**Comment by kgrzybek** [comment, marked as answer]
Hi @pawelpodsiadly,

I think we need to divide documentation into two areas:
1. Individual use case. In that case, my documentation is a set of tests (unit and integration) in AAA (Arrange, Act, Assert) / GWT (Given, When, Then) nomenclature. It is Executable Specification (BDD term), always synchronized and up to date.

2. System flow. This is harder to automate so at that moment I use the Event Storming technique and try to do it only for more complex flows. The drawback is that it is not synchronized with the code. There are 2 ways to solve this:
- using reflection and "diagram as code/text" (static analysis)
- generate it using data from working system (see https://docs.particular.net/serviceinsight/ for instance) (dynamic analysis)

Implementation of one of the solution for point 2 will be one of my future project I think.
